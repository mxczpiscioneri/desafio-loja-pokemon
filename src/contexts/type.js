import React, { createContext, useContext } from 'react'
import PropTypes from 'prop-types'
import { useLocalStorage } from '../hooks/useLocalStorage'

const TypeContext = createContext({})

export const TypeProvider = ({ children }) => {
  const { storedValue, setValue } = useLocalStorage()

  return (
    <TypeContext.Provider value={{ type: storedValue.type, setType: setValue }}>
      {children}
    </TypeContext.Provider>
  )
}

export const useTypeContext = () => useContext(TypeContext)

TypeProvider.propTypes = {
  children: PropTypes.any.isRequired
}

export default TypeContext
