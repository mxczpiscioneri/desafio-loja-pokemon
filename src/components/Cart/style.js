import styled from 'styled-components'
import { List, ListItem, ListSubheader } from '@material-ui/core'

export const ListStyled = styled(List)`
  width: 100%;
  max-width: 360px;
  background-color: #fafafa;
`

export const ListSubheaderStyled = styled(ListSubheader)`
  box-shadow: 0 1px 3px rgb(0 0 0 / 12%), 0 1px 2px rgb(0 0 0 / 24%);
  display: flex;
  align-items: center;
  font-weight: bold;
  font-size: 16px;

  svg {
    margin-right: 16px;
  }
`

export const ListItemResumeStyled = styled(ListItem)`
  background-color: #ccc;

  span {
    float: left;
  }

  p {
    float: right;
  }
`
