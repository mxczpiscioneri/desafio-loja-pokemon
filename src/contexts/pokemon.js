import React, { createContext, useContext } from 'react'
import PropTypes from 'prop-types'
import { keys, useLocalStorage } from '../hooks/useLocalStorage'

const PokemonContext = createContext({})

export const PokemonProvider = ({ children }) => {
  const { storedValue, setValue } = useLocalStorage()

  const setType = (type) => {
    setValue(keys.type, type)
  }

  const addCart = (pokemon) => {
    const listPokemon = JSON.parse(storedValue.cart)
    listPokemon.push(pokemon)

    setValue(keys.cart, JSON.stringify(listPokemon))
  }

  const removeCart = (index) => {
    const listPokemon = JSON.parse(storedValue.cart)
    const newListPokemon = listPokemon.filter((_, i) => i !== index)

    setValue(keys.cart, JSON.stringify(newListPokemon))
  }

  const clearCart = () => {
    setValue(keys.cart, JSON.stringify([]))
  }

  return (
    <PokemonContext.Provider value={{
      type: storedValue.type,
      setType,
      cart: JSON.parse(storedValue.cart),
      addCart,
      removeCart,
      clearCart }}
    >
      {children}
    </PokemonContext.Provider>
  )
}

export const usePokemonContext = () => useContext(PokemonContext)

PokemonProvider.propTypes = {
  children: PropTypes.any.isRequired
}

export default PokemonContext
