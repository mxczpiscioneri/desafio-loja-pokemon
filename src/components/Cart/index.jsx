import React, { Fragment } from 'react'
import Swal from 'sweetalert2'
import withReactContent from 'sweetalert2-react-content'
import ShoppingCartIcon from '@material-ui/icons/ShoppingCart'
import HighlightOffIcon from '@material-ui/icons/HighlightOff'
import { Avatar, Button, Divider, IconButton, ListItem, ListItemAvatar, ListItemSecondaryAction, ListItemText } from '@material-ui/core'
import { ListItemResumeStyled, ListStyled, ListSubheaderStyled } from './style'
import { usePokemonContext } from '../../contexts/pokemon'

const Cart = () => {
  const { cart, removeCart, clearCart } = usePokemonContext()
  let total = 0

  const handlerFinish = () => {
    const text = `Você escolheu ${cart.length} Pokemón.<br />O valor total é de $ ${total}.`
    const MySwal = withReactContent(Swal)
    MySwal.fire({
      title: 'Revisão do pedido',
      html: text,
      showCancelButton: true,
      confirmButtonText: 'Concluir pedido',
      cancelButtonText: 'Voltar',
      reverseButtons: true,
      allowOutsideClick: false
    }).then((result) => {
      if (result.isConfirmed) {
        clearCart()
        MySwal.fire({
          icon: 'success',
          title: 'Obrigado pela compra!',
          text: `Você ganhou $ ${total / 10} (10%) de cashback!`,
          showConfirmButton: false,
          timerProgressBar: true,
          timer: 5000
        })
      }
    })
  }

  if (!cart || cart.length === 0) {
    return <Fragment />
  }
  return (
    <>
      <ListStyled subheader={<ListSubheaderStyled><ShoppingCartIcon />Carrinho de compras</ListSubheaderStyled>}>
        {cart.map((pokemon, index) => {
          total += pokemon.base_experience
          return (
            <Fragment key={index}>
              <ListItem>
                <ListItemAvatar>
                  <Avatar src={pokemon.sprites.front_default} />
                </ListItemAvatar>
                <ListItemText primary={pokemon.name} secondary={`$ ${pokemon.base_experience}`} />
                <ListItemSecondaryAction>
                  <IconButton size="small" edge="end" aria-label="deletar" onClick={() => removeCart(index)}>
                    <HighlightOffIcon />
                  </IconButton>
                </ListItemSecondaryAction>
              </ListItem>
              <Divider />
            </Fragment>
          )
        })}
        <ListItemResumeStyled>
          <ListItemText primary='Total' secondary={`$ ${total}`} />
        </ListItemResumeStyled>
        <ListItemResumeStyled>
          <Button data-testid="btnFinish" size="large" variant="contained" color="primary" onClick={handlerFinish}>
            Continuar
          </Button>
        </ListItemResumeStyled>
      </ListStyled>
    </>
  )
}

export default Cart
