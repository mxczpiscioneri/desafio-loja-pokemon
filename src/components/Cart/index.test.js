import React from 'react'
import { act, fireEvent, render, screen } from '@testing-library/react'
import Cart from '.'
import * as PokemonContext from '../../contexts/pokemon'
import mockDetailsPokemon from '../../assets/mock-data/details-pokemon.json'

window.scrollTo = jest.fn()

describe('Cart components', () => {
  test('Render with success', () => {
    const mockContext = { cart: [mockDetailsPokemon] }
    jest
      .spyOn(PokemonContext, 'usePokemonContext')
      .mockImplementation(() => mockContext)

    const component = render(<Cart />)
    expect(component).toMatchSnapshot()
  })

  test('Render without pokemon', () => {
    const mockContext = { cart: [] }
    jest
      .spyOn(PokemonContext, 'usePokemonContext')
      .mockImplementation(() => mockContext)

    const component = render(<Cart />)
    expect(component).toMatchSnapshot()
  })

  test('remove pokemon', () => {
    const mockFunc = jest.fn()
    const mockContext = { cart: [mockDetailsPokemon], removeCart: mockFunc }
    jest
      .spyOn(PokemonContext, 'usePokemonContext')
      .mockImplementation(() => mockContext)

    render(<Cart />)

    fireEvent.click(screen.getByLabelText('deletar'))

    expect(mockFunc).toBeCalled()
  })

  test('clear cart', async () => {
    const mockFunc = jest.fn()
    const mockContext = { cart: [mockDetailsPokemon], clearCart: mockFunc }
    jest
      .spyOn(PokemonContext, 'usePokemonContext')
      .mockImplementation(() => mockContext)

    render(<Cart />)

    await act( async () =>{
      fireEvent.click(screen.getByTestId('btnFinish'))

      fireEvent.click(screen.getByText('Concluir pedido'))
    })

    expect(mockFunc).toBeCalled()
  })

  test('cancel checkout', () => {
    const mockFunc = jest.fn()
    const mockContext = { cart: [mockDetailsPokemon], clearCart: mockFunc }
    jest
      .spyOn(PokemonContext, 'usePokemonContext')
      .mockImplementation(() => mockContext)

    render(<Cart />)

    act(() =>{
      fireEvent.click(screen.getByTestId('btnFinish'))

      fireEvent.click(screen.getByText('Voltar'))
    })

    expect(mockFunc).not.toBeCalled()
  })
})
