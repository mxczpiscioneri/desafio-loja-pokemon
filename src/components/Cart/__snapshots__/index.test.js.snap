// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart components Render with success 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="MuiList-root sc-bdfBwQ laSfBL MuiList-padding MuiList-subheader"
      >
        <li
          class="MuiListSubheader-root sc-gsTCUz bLgCqt MuiListSubheader-sticky MuiListSubheader-gutters"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
          Carrinho de compras
        </li>
        <li
          class="MuiListItem-container"
        >
          <div
            class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle"
              >
                <img
                  class="MuiAvatar-img"
                  src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
                />
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                bulbasaur
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                $ 64
              </p>
            </div>
          </div>
          <div
            class="MuiListItemSecondaryAction-root"
          >
            <button
              aria-label="deletar"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
        <hr
          class="MuiDivider-root"
        />
        <li
          class="MuiListItem-root sc-dlfnbm kDiYaP MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Total
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              $ 64
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root sc-dlfnbm kDiYaP MuiListItem-gutters"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
            data-testid="btnFinish"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Continuar
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="MuiList-root sc-bdfBwQ laSfBL MuiList-padding MuiList-subheader"
    >
      <li
        class="MuiListSubheader-root sc-gsTCUz bLgCqt MuiListSubheader-sticky MuiListSubheader-gutters"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
        Carrinho de compras
      </li>
      <li
        class="MuiListItem-container"
      >
        <div
          class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
        >
          <div
            class="MuiListItemAvatar-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circle"
            >
              <img
                class="MuiAvatar-img"
                src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
              />
            </div>
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              bulbasaur
            </span>
            <p
              class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              $ 64
            </p>
          </div>
        </div>
        <div
          class="MuiListItemSecondaryAction-root"
        >
          <button
            aria-label="deletar"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14.59 8L12 10.59 9.41 8 8 9.41 10.59 12 8 14.59 9.41 16 12 13.41 14.59 16 16 14.59 13.41 12 16 9.41 14.59 8zM12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </li>
      <hr
        class="MuiDivider-root"
      />
      <li
        class="MuiListItem-root sc-dlfnbm kDiYaP MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Total
          </span>
          <p
            class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            $ 64
          </p>
        </div>
      </li>
      <li
        class="MuiListItem-root sc-dlfnbm kDiYaP MuiListItem-gutters"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          data-testid="btnFinish"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Continuar
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Cart components Render without pokemon 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
